   We have two problems when we start to use Xft1.2 or Xft2.0

   1)  With Xft1.2 or Xft2.0 we now use fontconfig, and thus have
       to cast the variables for use with fontconfig struct's and
       variables.

   2)  With Xft2.0, some functions in xftint.h are already defined
       in the Xft.h from Xft2.0.

   Fix:  If we have Xft1.2 or Xft2.0, either change configure to
         to define XFT_WITH_FONTCONFIG, or add -DXFT_WITH_FONTCONFIG
         to our CXXFLAGS.

         If we have Xft2.0, define HAVE_XFT2 somehow.


   Martin Schlemmer <azarah@gentoo.org> (24 Dec 2002)


--- kdebase-3.0.5a/kcontrol/kfontinst/kfontinst/XftConfig.cpp.orig	2002-10-30 00:53:18.000000000 +0200
+++ kdebase-3.0.5a/kcontrol/kfontinst/kfontinst/XftConfig.cpp	2002-10-30 01:01:49.000000000 +0200
@@ -276,7 +276,11 @@
             }
             case XftTypeString:
                 str+="\"";
+#ifndef XFT_WITH_FONTCONFIG
                 str+=test->value.u.s;
+#else
+                str+=(QCString &)test->value.u.s;
+#endif
                 str+="\"";
                 break;
             case XftTypeBool:
@@ -554,7 +554,11 @@
     XftValue           value;
     CXftConfig::TEntry *entry=new CXftConfig::TEntry;
 
+#ifndef XFT_WITH_FONTCONFIG
     value.u.s=sval;
+#else
+    (char *)value.u.s=sval;
+#endif
     value.type=XftTypeString;
     entry->test=XftTestCreate(XftQualAny, "family", XftOpEqual, value);
 
@@ -780,7 +780,11 @@
 
     for(entry=itsList.first(); entry; entry=itsList.next())
         if(entry->test && 0==CMisc::stricmp(entry->test->field, "family") && XftTypeString==entry->test->value.type &&
+#ifndef XFT_WITH_FONTCONFIG
            0==CMisc::stricmp(entry->test->value.u.s, family.latin1()) && entry->edit &&
+#else
+           0==CMisc::stricmp((char *)entry->test->value.u.s, family.latin1()) && entry->edit &&
+#endif
            0==CMisc::stricmp(entry->edit->field, field) && XftOpAssign==entry->edit->op && entry->edit->expr)
             break;
 
@@ -798,7 +798,11 @@
             free(test->field);
 
         if(XftTypeString==test->value.type)
+#ifndef XFT_WITH_FONTCONFIG
             free(test->value.u.s);
+#else
+            free((void *)test->value.u.s);
+#endif
     }
 }
 
--- kdebase-3.0.5a/kcontrol/kfontinst/kfontinst/XftConfigEditor.cpp.orig	2002-10-30 01:03:25.000000000 +0200
+++ kdebase-3.0.5a/kcontrol/kfontinst/kfontinst/XftConfigEditor.cpp	2002-10-30 01:07:27.000000000 +0200
@@ -469,7 +469,11 @@
                     break;
                 case XftTypeString:
                     expr->op=XftOpString;
+#ifndef XFT_WITH_FONTCONFIG
                     expr->u.sval=val.u.s;
+#else
+                    expr->u.sval=(char *)val.u.s;
+#endif
                     break;
                 case XftTypeBool:
                 default:
@@ -655,12 +655,24 @@
     if("foundry"==field || "family"==field || "encoding"==field || "style"==field)
     {
         val.type=XftTypeString;
+#ifndef XFT_WITH_FONTCONFIG
         val.u.s=(char *)malloc(edit ? strVal.length()+1 : strVal.length()-1);
+#else
+        (char *)val.u.s=(char *)malloc(edit ? strVal.length()+1 : strVal.length()-1);
+#endif
 
         if(edit)
+#ifndef XFT_WITH_FONTCONFIG
             strcpy(val.u.s, strVal.local8Bit());
+#else
+            strcpy((char *)val.u.s, strVal.local8Bit());
+#endif
         else
+#ifndef XFT_WITH_FONTCONFIG
             strcpy(val.u.s, strVal.mid(1, strVal.length()-2).local8Bit());  // Need to remove quotes from each end of the string...
+#else
+            strcpy((char *)val.u.s, strVal.mid(1, strVal.length()-2).local8Bit());  // Need to remove quotes from each end of the string...
+#endif
     }
     else if("spacing"==field)
     {
@@ -747,8 +747,13 @@
         {
             // Not sure about this...
             val.type=XftTypeString;
+#ifndef XFT_WITH_FONTCONFIG
             val.u.s=(char *)malloc(strVal.length()+1);
             strcpy(val.u.s, strVal.local8Bit());
+#else
+            (char *)val.u.s=(char *)malloc(strVal.length()+1);
+            strcpy((char *)val.u.s, strVal.local8Bit());
+#endif
         }
         else if("rgba"==field)
         {
--- kdebase-3.0.5a/kcontrol/kfontinst/kfontinst/xftint.h.orig	2002-12-24 23:34:28.000000000 +0200
+++ kdebase-3.0.5a/kcontrol/kfontinst/kfontinst/xftint.h	2002-12-24 23:35:22.000000000 +0200
@@ -276,8 +276,10 @@
 XftSubstPrint (XftSubst *subst);
 
 /* xftdir.c */
+#ifndef HAVE_XFT2
 Bool
 XftDirScan (XftFontSet *set, const char *dir);
+#endif
 
 /* xftdpy.c */
 int
@@ -388,11 +390,13 @@
 XftListValueCompare (XftValue	v1,
 		     XftValue	v2);
 
+#ifndef HAVE_XFT2
 Bool
 XftListValueListCompare (XftValueList	*v1orig,
 			 XftValueList	*v2orig,
 			 XftQual	qual);
+#endif
 
 Bool
 XftListMatch (XftPattern    *p,
 	      XftPattern    *font,
